hedera-node/hedera-network-admin-service-impl/src/main/java/com/hedera/node/app/service/networkadmin/impl/handlers/ReadableFreezeUpgradeActions.java

```diff
@@ -179,6 +201,87 @@ private void extractAndReplaceArtifacts(
         }
     }
 
+    private void generateConfigPem(@NonNull final Path artifactsLoc) {
+        requireNonNull(artifactsLoc, "Cannot generate config.txt without a valid artifacts location");
+
+        final var configTxt = artifactsLoc.resolve("config.txt");
+        final var nodeState = nodeStore.nodesState();
+
+        if (nodeState == null || nodeState.size() == 0) {
+            log.info("Node state is empty, cannot generate config.txt"); // change to log error later
+            return;
+        }
+
+        final var nodeIds = nodeStore.nodesState().keys();
+
+        try (FileWriter fw = new FileWriter(configTxt.toFile());
+                BufferedWriter bw = new BufferedWriter(fw)) {
+            Streams.stream(nodeIds)
+                    .map(nodeId -> nodeStore.get(nodeId.number()))
+                    .filter(node -> node != null && !node.deleted())
+                    .sorted(Comparator.comparing(Node::nodeId))
+                    .forEach(node -> writeConfigLineAndPem(node, bw, artifactsLoc));
+        } catch (final IOException e) {
+            log.error("Failed to generate {} with exception : {}", configTxt, e);
+        }
+    }
+
+    private void writeConfigLineAndPem(
+            @NonNull final Node node, @NonNull final BufferedWriter bw, @NonNull final Path pathToWrite) {
+        requireNonNull(node);
+        requireNonNull(bw);
+        requireNonNull(pathToWrite);
+
+        var line = new StringBuilder();
+        int weight = 0;
+        final var name = "node" + node.nodeId();
+        final var alias = nameToAlias(name);
+        final var pemFile = pathToWrite.resolve("s-public-" + alias + ".pem");
+        final int INT = 0;
+        final int EXT = 1;
+
+        final var stakingNodeInfo = stakingInfoStore.get(node.nodeId());
+        if (stakingNodeInfo != null) {
```

Comment:
Can this ever be null ? We should add `requireNonNull` here

Reply:
If a new node is added and StakingInfoStore has not get time to be updated with the new node info (such as before midnight), and we started freeze upgrade, we could get stakingNodeInfo = null here. Then we just use default weight 0 here.

Reply:
Ah make sense. Thanks @iwsimon